pkg/ingester/limiter.go

```diff
@@ -173,15 +176,23 @@ func convertGlobalToLocalLimit(globalLimit int, ringCount RingCount, replication
 	// topology changes) and we prefer to always be in favor of the tenant,
 	// we can use a per-ingester limit equal to:
 	// (global limit / number of ingesters) * replication factor
-	numIngesters := ringCount.HealthyInstancesCount()
+	numIngesters := l.ring.HealthyInstancesCount()
 
 	// May happen because the number of ingesters is asynchronously updated.
 	// If happens, we just temporarily ignore the global limit.
 	if numIngesters > 0 {
-		return int((float64(globalLimit) / float64(numIngesters)) * float64(replicationFactor))
+		return int((float64(globalLimit) / float64(numIngesters)) * float64(l.replicationFactor))
 	}
 
-	return 0
+	// If the number of available ingesters is greater than the tenant's shard
+	// size, then we should honor the shard size because series/metadata won't
+	// be written to more ingesters than it.
+	if shardSize := l.limits.IngestionTenantShardSize(tenantID); shardSize > 0 {
+		// We use Min() to protect from the case the expected shard size is > available ingesters.
+		numIngesters = lo.Min([]int{numIngesters, util.ShuffleShardExpectedInstances(shardSize, 1)})
+	}
+
+	return int((float64(globalLimit) / float64(numIngesters)) * float64(l.replicationFactor))
```

Comment:
Good spot, we really need a test to cover the limit creation.

I thought something along the lines "already" there, because the lines looked copied in from mimir :joy:  :-1: 
https://github.com/grafana/mimir/blob/3a441e99785ef49fda246ef65326143765360b74/pkg/ingester/limiter.go#L193-L197

Reply:
Agreed!